packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,493 @@
+  typeof controllerName,
+  MultichainAssetsControllerActions | AllowedActions,
+  MultichainAssetsControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
+
+/**
+ * {@link MultichainAssetsController}'s metadata.
+ *
+ * This allows us to choose if fields of the state should be persisted or not
+ * using the `persist` flag; and if they can be sent to Sentry or not, using
+ * the `anonymous` flag.
+ */
+const assetsControllerMetadata = {
+  metadata: {
+    persist: true,
+    anonymous: false,
+  },
+  allNonEvmTokens: {
+    persist: true,
+    anonymous: false,
+  },
+};
+
+// Define a temporary interface for the permission structure
+type AssetEndowment = {
+  'endowment:assets'?: {
+    scopes: CaipChainId[];
+  };
+};
+
+// TODO make this controller extends StaticIntervalPollingController and update all metadata once a day.
+
+export class MultichainAssetsController extends BaseController<
+  typeof controllerName,
+  MultichainAssetsControllerState,
+  MultichainAssetsControllerMessenger
+> {
+  // Mapping of CAIP-2 Chain ID to Asset Snaps.
+  #snaps: Record<CaipChainId, Snap[]>;
+
+  constructor({
+    messenger,
+    state = {},
+  }: {
+    messenger: MultichainAssetsControllerMessenger;
+    state?: Partial<MultichainAssetsControllerState>;
+  }) {
+    super({
+      messenger,
+      name: controllerName,
+      metadata: assetsControllerMetadata,
+      state: {
+        ...getDefaultMultichainAssetsControllerState(),
+        ...state,
+      },
+    });
+
+    this.#snaps = {};
+
+    this.messagingSystem.subscribe(
+      'AccountsController:accountAdded',
+      async (account) => await this.#handleOnAccountAdded(account),
+    );
+    this.messagingSystem.subscribe(
+      'AccountsController:accountRemoved',
+      async (account) => await this.#handleOnAccountRemoved(account),
+    );
+  }
+
+  /**
+   * Function to update the assets list for an account
+   *
+   * @param list - The list of assets to update
+   */
+  async updateAccountAssetsList(list: AccountAssetListUpdatedEvent) {
+    const assetsToUpdate = list.params.assets;
+    for (const accountId in assetsToUpdate) {
+      if (Object.prototype.hasOwnProperty.call(assetsToUpdate, accountId)) {
+        const newAccountAssets = assetsToUpdate[accountId];
+        const assets = this.state.allNonEvmTokens[accountId] || [];
+
+        const filteredAssetsToAdd = newAccountAssets.added.filter(
+          (asset) => !assets.includes(asset),
+        );
+        // trigger fetching metadata for new assets
+        await this.#refreshAssetsSnapsFor(filteredAssetsToAdd);
+        const newAssets = [...assets, ...filteredAssetsToAdd];
+
+        const assetsAfterRemoval = newAssets.filter(
+          (asset) => !newAccountAssets.removed.includes(asset),
+        );
+        this.update((state) => {
+          state.allNonEvmTokens[accountId] = assetsAfterRemoval;
+        });
```

Comment:
could this be done in a away that
* You update assets
* Then take metadata for all assets
Or the other way around 
Instead of taking the metadata each account? Should be a faster iteration

Reply:
That makes sense, i can update metadata at the end for all assets

Reply:
Updated logic here [20ef8da](https://github.com/MetaMask/core/pull/5138/commits/20ef8da738a5a088191cf323d4a343f41141c85d)